gc()

#####THIS PIECE OF CODE USES RESULTS GENERATED BY fit_JLS.R

library(data.table)
library(lassosum) #transform p value to correlation
library(doParallel) # foreach
library(R.utils)
# library(pROC) # for AUC 
# library(pryr) # check memory useage
JLS_population_weight <- c(0, 0.5, 1) #gamma parameter in the paper
JLS_population_weight_one <- JLS_population_weight[2]

JLS_result_prefix <- '/raid6/Tianyu/PRS/sharable/result/JLS_result_weight_is'
small_population_reference_prefix_merged <- '/raid6/Tianyu/PRS/sharable/data/YRI-chr21n22'

###THE JLS MODELS ARE STORED SEPARATELY FOR DIFFERENT POPULATION WEIGHTS
JLS_regression_coefficient_file <- paste0(JLS_result_prefix, 
                                          sprintf("%.2f",JLS_population_weight_one), '_coefficient.txt')
JLS_regression_coefficient <- fread(JLS_regression_coefficient_file)

JLS_result_one_weight_file <- paste0(JLS_result_prefix, 
                                     sprintf("%.2f",JLS_population_weight_one), '.Rdata')

JLS_result_one_weight <- get(load(JLS_result_one_weight_file))

for(l1_penalty_index in 1:length(JLS_result_one_weight$lambda)){
  # column_name_of_coefficient <- paste0('l1_para_is_',JLS_result_one_weight$lambda[l1_penalty_index])
  effect_size_df <- data.frame(SNP = JLS_regression_coefficient$ID, 
                               A1 = unlist(lapply(strsplit(JLS_regression_coefficient$ID, ":"),`[[`,4)),
                               BETA = JLS_regression_coefficient[, ..l1_penalty_index])
  
  selected_coefficient_file <- paste0(JLS_result_prefix, 
                                      sprintf("%.2f",JLS_population_weight_one),
                                      '_l1_penalty_is_',
                                      sprintf("%.4f",JLS_result_one_weight$lambda[l1_penalty_index]), '_coefficient.txt')
  write.table(effect_size_df, selected_coefficient_file, sep = "\t", quote = FALSE, row.names = FALSE)
  
  PGS_file <- paste0(JLS_result_prefix, 
                     sprintf("%.2f",JLS_population_weight_one),
                     '_l1_penalty_is_',
                     sprintf("%.4f",JLS_result_one_weight$lambda[l1_penalty_index]), '_PGS')
  
  plink2.command = paste("plink2 --nonfounders","--allow-no-sex","--threads", 8,"--memory", 25000,
                         "--bfile", small_population_reference_prefix_merged ,
                         "--score", selected_coefficient_file, "header-read",1,2,
                         "--score-col-nums",3,
                         "--out",PGS_file,
                         sep=" ")
  
  system(plink2.command)
}


###I USE THE FOLLOWING LINES TO MERGE THE CHR21 CHR22 PLINK FILES TOGETHER
# small_population_reference_prefix_21 <- '/raid6/Tianyu/PRS/sharable/data/YRI-chr21'
# small_population_reference_prefix_22 <- '/raid6/Tianyu/PRS/sharable/data/YRI-chr22'
# plink_merge_command <- paste0("plink --bfile ", small_population_reference_prefix_21,
#                               " --bmerge ", small_population_reference_prefix_22, ".bed ",
#                               small_population_reference_prefix_22, ".bim ", 
#                               small_population_reference_prefix_22, ".fam ",
#                               "--out ", small_population_reference_prefix_merged)
# system(plink_merge_command)



##########read in lasso results##########
TrainJLFile <- paste0(work.dir, 
                      'JointLassoSum/JointLassosum--gamma-', 
                      sprintf("%.2f",gammaGenerateData), 
                      '.Rdata')
TrainJLResult <- get(load(TrainJLFile))
AllBeta <- TrainJLResult$beta

betaGenerateData <- AllBeta[, lambdaIndexGenerateData] #use a small lambda to generate bootstrap data
##########read in genotype and calculate score#############
map <- fread(paste0(main.dir,
                    'Data/chr1-22-qc-frq-ld.block.map'))
CHR <- gsub("chr","",map$CHROM)
SNP <- map$ID
names(betaGenerateData) <- SNP


# ######SECTION 1: calculate PGS of the new individuals####
# risk.score.list <- vector("list",2)
####save the selected preliminary beta
beta_generate_data_file <- './temp_file_generate_data/beta_generate_data.txt'
A1 <- unlist(lapply(strsplit(SNP, ":"),`[[`,4))
effect_size_df <- data.frame(SNP = SNP, 
                             A1 = A1,
                             BETA = betaGenerateData)
write.table(effect_size_df, beta_generate_data_file, sep = "\t", quote = FALSE, row.names = FALSE)

ancs <- c('CEU', 'YRI')
for(i.set in 1:2){
  anc <- ancs[i.set]
  # re.pgss <- mclapply(chrs, PGS_by_chr, anc = anc,
  #                     beta0 = betaGenerateData, mc.cores = 8, mc.preschedule = FALSE)
  # 
  # 
  # ### sum the results
  # pgs <- re.pgss[[1]] #this is the first chromosome
  # for(i in 2:length(re.pgss)){
  #   pgs <- pgs+re.pgss[[i]]
  # }
  # 
  # risk.score.list[[i.set]] <- pgs
  
  ##calculate the PRS for each individual in the reference panel
  if(anc == 'CEU'){
    file.title <- 'CEU-20K'
  }else{
    file.title <- 'YRI-4K'  
  }
  
  
}
### save risk scores for each population

# save(risk.score.list, file = paste0(ParameterTuningDirectory,
# '/riskscore.Rdata'))





source('JLS_function.R')
source('JLS_function_supp.R')
Rcpp::sourceCpp("JLS_function.cpp")

### FILE LOCATION
large_population_GWAS_file <- '/raid6/Tianyu/PRS/sharable/data/large_population_GWAS_two_chr'
small_population_GWAS_file <- '/raid6/Tianyu/PRS/sharable/data/small_population_GWAS_two_chr'
large_population_reference_prefix <- '/raid6/Tianyu/PRS/sharable/data/CEU-chr'


JLS_result_prefix <- '/raid6/Tianyu/PRS/sharable/result/JLS_result_weight_is'
###OTHER METADATA
large_population_type <- 'CEU' #or EUR, for determining LD block boundaries
small_population_type <- 'YRI' #or AFR
# small_population_type <- 'ASN'

###HYPERPARAMETER CANDIDATES
JLS_population_weight <- c(0, 0.5, 1) #gamma parameter in the paper
# JLS_l1_penalty <- exp(seq(log(0.007), log(0.05), length.out=5))
JLS_l1_penalty <- exp(seq(log(0.01), log(0.005), length.out = 2)) ##lambda
JLS_shrinkage <- c(0.9) #s
chromosome <- 21:22 ###usually you need to change this to 1:22 

###GIVEN THE ABOVE INFORMATION, FIT THE MODEL FOR ONE TIME
# JLS_population_weight_one <- JLS_population_weight[1]

#####BASIC USAGE FOR ONE GIVEN POPULATION WEIGHT PARAMETER, THE OTHER TWO PARAMETERS CAN BE VECTORS
# JLS_train(JLS_population_weight_one = JLS_population_weight_one,
#           JLS_l1_penalty = JLS_l1_penalty,
#           JLS_shrinkage = JLS_shrinkage,
#           large_population_GWAS_file = large_population_GWAS_file,
#           small_population_GWAS_file = small_population_GWAS_file,
#           large_population_reference_prefix = large_population_reference_prefix,
#           small_population_reference_prefix = small_population_reference_prefix,
#           JLS_result_prefix = JLS_result_prefix,
#           large_population_type = large_population_type,
#           small_population_type = small_population_type,
#           chromosome = chromosome)

##PARALLEL OVER DIFFERENT WEIGHT PARAMETER GAMMA
mclapply(JLS_population_weight,
         FUN = JLS_train,
         JLS_l1_penalty = JLS_l1_penalty,
         JLS_shrinkage = JLS_shrinkage,
         large_population_GWAS_file = large_population_GWAS_file,
         small_population_GWAS_file = small_population_GWAS_file,
         large_population_reference_prefix = large_population_reference_prefix,
         small_population_reference_prefix = small_population_reference_prefix,
         JLS_result_prefix = JLS_result_prefix,
         large_population_type = large_population_type,
         small_population_type = small_population_type,
         chromosome = chromosome,
         mc.cores = 3, 
         mc.preschedule = F, 
         mc.silent = F)

####NOW THE FITTING RESULTS ARE IN THE SPECIFIED LOCATION

# large_population_GWAS_two_chr <- large_population_GWAS[large_population_GWAS$`#CHROM` %in% c(21,22), ]
# small_population_GWAS_two_chr <- small_population_GWAS[small_population_GWAS$`#CHROM` %in% c(21,22), ]
# large_population_GWAS_two_chr_file <- '/raid6/Tianyu/PRS/sharable/large_population_GWAS_two_chr'
# small_population_GWAS_two_chr_file <- '/raid6/Tianyu/PRS/sharable/small_population_GWAS_two_chr'
# fwrite(large_population_GWAS_two_chr, file = large_population_GWAS_two_chr_file)
# fwrite(small_population_GWAS_two_chr, file = small_population_GWAS_two_chr_file)


#

# source("Lassosum_function/parseselect.R")
# source("Lassosum_function/parseblocks.R")
# source("Lassosum_function/ncol.bfile.R")
# source("Lassosum_function/nrow.bfile.R")
# source("Lassosum_function/read.table2.R")
# source("Lassosum_function/selectregion.R")
# source("Lassosum_function/parse.pheno.covar.R")
# source("Lassosum_function/splitgenome.R")
# source("Lassosum_function/validation.R")
# source("Lassosum_function/merge.mylassosum.R")

# large_population_GWAS_file <- '/raid6/Ron/prs/data/bert_sample/CEU.TRN.PHENO1.glm.logistic.hybrid'
# small_population_GWAS_file <- '/raid6/Ron/prs/data/bert_sample/YRI.TRN.PHENO1.glm.logistic.hybrid'
# large_population_reference_prefix <- '/raid6/Tianyu/PRS/SimulationPipeline/Data/Reference-LDblocks/CEU/CHR/CEU-chr'
# small_population_reference_prefix <- '/raid6/Tianyu/PRS/SimulationPipeline/Data/Reference-LDblocks/YRI/CHR/YRI-chr'


# large_population_GWAS_file <- '/Users/tianyuzhang/Documents/GitHub/JointLassosum/JLS_basic/data/large_population_GWAS_two_chr'
# small_population_GWAS_file <- '/Users/tianyuzhang/Documents/GitHub/JointLassosum/JLS_basic/data/small_population_GWAS_two_chr'
# large_population_reference_prefix <- '/Users/tianyuzhang/Documents/GitHub/JointLassosum/JLS_basic/data/CEU-chr'
# small_population_reference_prefix <- '/Users/tianyuzhang/Documents/GitHub/JointLassosum/JLS_basic/data/YRI-chr'
# gene_ID_chromosome_file <- '/raid6/Ron/prs/data/bert_sample/CEU.TRN/CEU.TRN.pvar'

